I will be working with multiple data such as housing prices, population and migration.

Import all the necessary packages first.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

I'd like to start with housing prices dataset that I downloaded from Zillow website.

Loading data from csv file.

In [2]:
homes_url = 'Data/City_Zhvi_AllHomes.csv'
homes = pd.read_csv(homes_url, encoding='windows-1252')

Let's see how big is the dataset.

In [3]:
homes.shape

(12933, 278)

It has 12933 rows and 278 columns. Now let's see the column names.

In [4]:
homes.columns

Index(['RegionID', 'RegionName', 'State', 'Metro', 'CountyName', 'SizeRank',
       '1996-04', '1996-05', '1996-06', '1996-07',
       ...
       '2018-02', '2018-03', '2018-04', '2018-05', '2018-06', '2018-07',
       '2018-08', '2018-09', '2018-10', '2018-11'],
      dtype='object', length=278)

Ok. I can see that this dataset contains monthly average prices starting from April 1996 until November 2018. I have 22 years of data. Wow!

My next question what are the data types of each column?

In [5]:
homes.dtypes

RegionID        int64
RegionName     object
State          object
Metro          object
CountyName     object
SizeRank        int64
1996-04       float64
1996-05       float64
1996-06       float64
1996-07       float64
1996-08       float64
1996-09       float64
1996-10       float64
1996-11       float64
1996-12       float64
1997-01       float64
1997-02       float64
1997-03       float64
1997-04       float64
1997-05       float64
1997-06       float64
1997-07       float64
1997-08       float64
1997-09       float64
1997-10       float64
1997-11       float64
1997-12       float64
1998-01       float64
1998-02       float64
1998-03       float64
               ...   
2016-06         int64
2016-07         int64
2016-08         int64
2016-09         int64
2016-10         int64
2016-11         int64
2016-12         int64
2017-01         int64
2017-02         int64
2017-03         int64
2017-04         int64
2017-05         int64
2017-06         int64
2017-07         int64
2017-08   

From this I can conclude that some price columns are float and others are integers. I need to conform to one type, which I will do later.

Let's take a look at the data.

In [6]:
homes.head()

RegionID   RegionName State                             Metro  \
0      6181     New York    NY       New York-Newark-Jersey City   
1     12447  Los Angeles    CA    Los Angeles-Long Beach-Anaheim   
2     17426      Chicago    IL          Chicago-Naperville-Elgin   
3     39051      Houston    TX  Houston-The Woodlands-Sugar Land   
4      6915  San Antonio    TX         San Antonio-New Braunfels   

           CountyName  SizeRank   1996-04   1996-05   1996-06   1996-07  \
0       Queens County         1  169800.0  170100.0  170500.0  170700.0   
1  Los Angeles County         2  158800.0  158400.0  158100.0  157600.0   
2         Cook County         3  112800.0  114200.0  115400.0  116200.0   
3       Harris County         4   86500.0   86600.0   86600.0   86700.0   
4        Bexar County         5   86600.0   86400.0   86300.0   86300.0   

    ...     2018-02  2018-03  2018-04  2018-05  2018-06  2018-07  2018-08  \
0   ...      652500   657000   661600   666500   669600   670500   670500   
1   ...      664700   669300   671300   673300   674700   675800   677600   
2   ...      227300   228300   228700   228500   228100   227700   227700   
3   ...      174500   175700   176700   177800   178800   179900   180900   
4   ...      164000   165600   166600   167300   168100   169100   170100   

   2018-09  2018-10  2018-11  
0   672700   676900   680400  
1   680100   683200   686500  
2   227900   228000   228500  
3   181500   182100   183000  
4   171300   172500   173500  

[5 rows x 278 columns]

In order to make my work easier I will split this dataset into two separate dataframes. The first will contain the description, and the second one will have only numeric data of prices by months and years.

In [7]:
homes_info = homes[['RegionID', 'RegionName', 'State', 'Metro', 'CountyName', 'SizeRank']]
homes_info.head()

RegionID   RegionName State                             Metro  \
0      6181     New York    NY       New York-Newark-Jersey City   
1     12447  Los Angeles    CA    Los Angeles-Long Beach-Anaheim   
2     17426      Chicago    IL          Chicago-Naperville-Elgin   
3     39051      Houston    TX  Houston-The Woodlands-Sugar Land   
4      6915  San Antonio    TX         San Antonio-New Braunfels   

           CountyName  SizeRank  
0       Queens County         1  
1  Los Angeles County         2  
2         Cook County         3  
3       Harris County         4  
4        Bexar County         5

In [8]:
homes_prices = homes.iloc[:, 6:]
homes_prices.head()

1996-04   1996-05   1996-06   1996-07   1996-08   1996-09   1996-10  \
0  169800.0  170100.0  170500.0  170700.0  170800.0  170800.0  170700.0   
1  158800.0  158400.0  158100.0  157600.0  157300.0  157000.0  156800.0   
2  112800.0  114200.0  115400.0  116200.0  116600.0  116500.0  116000.0   
3   86500.0   86600.0   86600.0   86700.0   87000.0   87300.0   87600.0   
4   86600.0   86400.0   86300.0   86300.0   86200.0   86100.0   86000.0   

    1996-11   1996-12   1997-01   ...     2018-02  2018-03  2018-04  2018-05  \
0  170700.0  170800.0  171100.0   ...      652500   657000   661600   666500   
1  156900.0  157200.0  157600.0   ...      664700   669300   671300   673300   
2  115100.0  114300.0  113600.0   ...      227300   228300   228700   228500   
3   87800.0   87900.0   87800.0   ...      174500   175700   176700   177800   
4   86000.0   85900.0   85900.0   ...      164000   165600   166600   167300   

   2018-06  2018-07  2018-08  2018-09  2018-10  2018-11  
0   669600   670500   670500   672700   676900   680400  
1   674700   675800   677600   680100   683200   686500  
2   228100   227700   227700   227900   228000   228500  
3   178800   179900   180900   181500   182100   183000  
4   168100   169100   170100   171300   172500   173500  

[5 rows x 272 columns]

Once I have separated prices, I can now turn integers into float as I think for the purposes of my project it is better to work with float. I also need to check the dataframe for empty values. I will do it across the columns.

In [9]:
homes_prices.astype(float).head()

1996-04   1996-05   1996-06   1996-07   1996-08   1996-09   1996-10  \
0  169800.0  170100.0  170500.0  170700.0  170800.0  170800.0  170700.0   
1  158800.0  158400.0  158100.0  157600.0  157300.0  157000.0  156800.0   
2  112800.0  114200.0  115400.0  116200.0  116600.0  116500.0  116000.0   
3   86500.0   86600.0   86600.0   86700.0   87000.0   87300.0   87600.0   
4   86600.0   86400.0   86300.0   86300.0   86200.0   86100.0   86000.0   

    1996-11   1996-12   1997-01    ...      2018-02   2018-03   2018-04  \
0  170700.0  170800.0  171100.0    ...     652500.0  657000.0  661600.0   
1  156900.0  157200.0  157600.0    ...     664700.0  669300.0  671300.0   
2  115100.0  114300.0  113600.0    ...     227300.0  228300.0  228700.0   
3   87800.0   87900.0   87800.0    ...     174500.0  175700.0  176700.0   
4   86000.0   85900.0   85900.0    ...     164000.0  165600.0  166600.0   

    2018-05   2018-06   2018-07   2018-08   2018-09   2018-10   2018-11  
0  666500.0  669600.0  670500.0  670500.0  672700.0  676900.0  680400.0  
1  673300.0  674700.0  675800.0  677600.0  680100.0  683200.0  686500.0  
2  228500.0  228100.0  227700.0  227700.0  227900.0  228000.0  228500.0  
3  177800.0  178800.0  179900.0  180900.0  181500.0  182100.0  183000.0  
4  167300.0  168100.0  169100.0  170100.0  171300.0  172500.0  173500.0  

[5 rows x 272 columns]

In [10]:
homes_prices.isnull().any()

1996-04     True
1996-05     True
1996-06     True
1996-07     True
1996-08     True
1996-09     True
1996-10     True
1996-11     True
1996-12     True
1997-01     True
1997-02     True
1997-03     True
1997-04     True
1997-05     True
1997-06     True
1997-07     True
1997-08     True
1997-09     True
1997-10     True
1997-11     True
1997-12     True
1998-01     True
1998-02     True
1998-03     True
1998-04     True
1998-05     True
1998-06     True
1998-07     True
1998-08     True
1998-09     True
           ...  
2016-06    False
2016-07    False
2016-08    False
2016-09    False
2016-10    False
2016-11    False
2016-12    False
2017-01    False
2017-02    False
2017-03    False
2017-04    False
2017-05    False
2017-06    False
2017-07    False
2017-08    False
2017-09    False
2017-10    False
2017-11    False
2017-12    False
2018-01    False
2018-02    False
2018-03    False
2018-04    False
2018-05    False
2018-06    False
2018-07    False
2018-08    False
2018-09    Fal

As a result, I can see that there are columns with NaN values. Let's see how many columns have missing values and what is the total number.

In [11]:
homes_prices.isnull().any().sum()

225

In [12]:
homes_prices.isnull().values.sum()

158225

Now I have a dilemma. How can I manage null values? Drop them or fill them with numbers? If Dropping is not an option as 225 columns out of 272 contain null values. Filling them is a big question here. How to fill? What numbers to fill with?

In [13]:
# move the following down below:
population_url = 'Data/1000-largest-us-cities-by-population-with-geographic-coordinates.json'
population = pd.read_json(population_url)

In [14]:
population.columns

Index(['datasetid', 'fields', 'geometry', 'record_timestamp', 'recordid'], dtype='object')

In [15]:
population['fields']

0      {'city': 'South San Francisco', 'rank': 522, '...
1      {'city': 'Aliso Viejo', 'rank': 741, 'state': ...
2      {'city': 'Rapid City', 'rank': 479, 'state': '...
3      {'city': 'Coon Rapids', 'rank': 563, 'state': ...
4      {'city': 'Malden', 'rank': 585, 'state': 'Mass...
5      {'city': 'Noblesville', 'rank': 644, 'state': ...
6      {'city': 'Delray Beach', 'rank': 543, 'state':...
7      {'city': 'Wellington', 'rank': 588, 'state': '...
8      {'city': 'Florissant', 'rank': 703, 'state': '...
9      {'city': 'Youngstown', 'rank': 532, 'state': '...
10     {'city': 'Newport Beach', 'rank': 356, 'state'...
11     {'city': 'Bloomington', 'rank': 412, 'state': ...
12     {'city': 'Menifee', 'rank': 384, 'state': 'Cal...
13     {'city': 'Lynchburg', 'rank': 423, 'state': 'V...
14     {'city': 'Loveland', 'rank': 472, 'state': 'Co...
15     {'city': 'St. Joseph', 'rank': 429, 'state': '...
16     {'city': 'Folsom', 'rank': 458, 'state': 'Cali...
17     {'city': 'Decatur', 'ran

In [16]:
population_list = population['fields'].tolist()

In [17]:
population = pd.DataFrame(population_list)
population.head()

city                 coordinates  growth_from_2000_to_2013  \
0  South San Francisco   [37.654656, -122.4077498]                       9.1   
1          Aliso Viejo  [33.5676842, -117.7256083]                      25.4   
2           Rapid City  [44.0805434, -103.2310149]                      17.9   
3          Coon Rapids   [45.1732394, -93.3030063]                       0.6   
4               Malden    [42.4250964, -71.066163]                       7.4   

   population  rank          state  
0       66174   522     California  
1       50175   741     California  
2       70812   479   South Dakota  
3       62103   563      Minnesota  
4       60509   585  Massachusetts

In [18]:
population = population[['city', 'state', 'population']]
population.sort_values(by='population', ascending=False)

city                 state  population
67          New York              New York     8405837
591      Los Angeles            California     3884307
602          Chicago              Illinois     2718782
70           Houston                 Texas     2195914
674     Philadelphia          Pennsylvania     1553165
855          Phoenix               Arizona     1513367
469      San Antonio                 Texas     1409019
536        San Diego            California     1355896
842           Dallas                 Texas     1257676
55          San Jose            California      998537
800           Austin                 Texas      885400
421     Indianapolis               Indiana      843393
115     Jacksonville               Florida      842583
28     San Francisco            California      837442
798         Columbus                  Ohio      822553
492        Charlotte        North Carolina      792862
247       Fort Worth                 Texas      792727
603          Detroit              Michigan      688701
69           El Paso                 Texas      674433
795          Memphis             Tennessee      653450
838          Seattle            Washington      652405
65            Denver              Colorado      649495
240       Washington  District of Columbia      646449
935           Boston         Massachusetts      645966
656        Nashville             Tennessee      634464
595        Baltimore              Maryland      622104
112    Oklahoma City              Oklahoma      610613
281       Louisville              Kentucky      609893
797         Portland                Oregon      609456
839        Las Vegas                Nevada      603488
..               ...                   ...         ...
271       Brookfield             Wisconsin       37999
892       Park Ridge              Illinois       37839
529         Florence        South Carolina       37792
588              Roy                  Utah       37733
491    Winter Garden               Florida       37711
530          Chelsea         Massachusetts       37670
687    Valley Stream              New York       37659
130      Spartanburg        South Carolina       37647
977      Lake Oswego                Oregon       37610
219      Friendswood                 Texas       37587
703      Westerville                  Ohio       37530
341       Northglenn              Colorado       37499
672      Phenix City               Alabama       37498
893       Grove City                  Ohio       37490
460        Texarkana                 Texas       37442
833          Addison              Illinois       37385
704            Dover              Delaware       37366
187     Lincoln Park              Michigan       37313
100     Calumet City              Illinois       37240
396         Muskegon              Michigan       37213
771         Aventura               Florida       37199
561         Martinez            California       37165
832       Greenfield             Wisconsin       37159
456  Apache Junction               Arizona       37130
959         Monrovia            California       37101
773          Weslaco                 Texas       37093
101           Keizer                Oregon       37064
398     Spanish Fork                  Utah       36956
350           Beloit             Wisconsin       36888
746      Panama City               Florida       36877

[1000 rows x 3 columns]

In [19]:
state_abbr = pd.read_csv('Data/state_abbr.csv')
state_abbr.head()
state_abbr.rename ************

SyntaxError: invalid syntax (<ipython-input-19-46663e3278ec>, line 3)

In [ ]:
homes1 = pd.merge(homes_info, state_abbr, how='left', left_on='State', right_on='Abbreviation')

In [ ]:
homes1.tail(25)

In [ ]:
homes_new = pd.merge(homes1, population, how='left', left_on=['RegionName', 'Abbreviation'], right_on=['city', 'state'])
homes_new.head()

In [ ]:
homes[homes['RegionName'] == 'Mountain View']

In [ ]:
prices = homes.drop(['RegionID', 'RegionName', 'State', 'Metro', 'CountyName', 'SizeRank'], axis=1)
prices.head()

In [ ]:
prices.isnull()

In [ ]:
prices = prices.fillna(axis=1, method='ffill')
prices = prices.fillna(axis=1, method='bfill')

In [ ]:
homes2 = pd.melt(homes, value_vars=prices, id_vars=['RegionID', 'RegionName', 'State'], var_name='Date', value_name='Price')


In [ ]:
homes2.head()

In [ ]:
prices.diff(periods=12, axis=1)